core/src/com/unciv/logic/automation/unit/CivilianUnitAutomation.kt

```diff
@@ -117,6 +123,33 @@ object CivilianUnitAutomation {
                 UnitActions.invokeUnitAction(unit, UnitActionType.StartGoldenAge)
         }
 
+        if (unit.hasUnique(UniqueType.GainFreeBuildings)) {
+            val unique = unit.getMatchingUniques(UniqueType.GainFreeBuildings).first()
+            val buildingName = unique.params[0]
+            // Choose the city that is closest in distance and does not have the building constructed.
+            val cityToGainBuilding = unit.civ.cities.filter {
+                !it.cityConstructions.containsBuildingOrEquivalent(buildingName)
+                    && (unit.movement.canMoveTo(it.getCenterTile()) || unit.currentTile == it.getCenterTile())
+            }.minByOrNull {
+                val path = unit.movement.getShortestPath(it.getCenterTile())
+                path.size
+            }
+
+            if (cityToGainBuilding != null) {
+                if (unit.currentTile == cityToGainBuilding.getCenterTile()) {
+                    UniqueTriggerActivation.triggerUnique(unique, unit.civ, unit = unit, tile = unit.currentTile)
+                    UnitActionModifiers.activateSideEffects(unit, unique)
```

Comment:
Instead, we should activate the unit action, which does all this

Reply:
> Instead, we should activate the unit action, which does all this

You mean replace it with the code `UnitActions.invokeUnitAction(unit, UnitActionType.TriggerUnique)` ?
But this would again only activate the **first** unit action, not necessarily the gain building action.